### ISSUE TITLE
dart process with +18GB of memory when CRTL + SPACE

### ISSUE THREAD

**Issue by peekpt** [flutter/flutter#23155]
Using VSCode

1. starts the debugging, physical or virtual device (doesn't matter)
 - debugging started, app runs, everything works.
2.  Edits code and hit CRTL+ SPACE to get the helpers
 - Loading... box appears nothing shows, hard disk starts to rumble hard 
 - dart process starts increasing memory
3. Hit several times CRTL + SPACE
 - memory increases hard, computer unstable, dart process with  18GB memory.

4. Runs $ killall dart
 - peace restored...
 - popup message shows in VSCode with a button to reload dart.
5. Reloads
 - 

CTRL+SPACE now works

![memory](https://user-images.githubusercontent.com/222385/47020886-3914fb80-d152-11e8-923e-1b033d26bf85.png)

---
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel dev, v0.10.0, on Mac OS X 10.13.6 17G65, locale pt-PT)
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 10.0)
[✓] Android Studio (version 3.1)
[✓] VS Code (version 1.28.1)
[✓] Connected device (1 available)

• No issues found!

**Comment by zoechi** [comment]
@DanTup any ideas?

**Comment by DanTup** [comment]
Yikes!

@peekpt Can you find what the command line was used to launch that process? Does it reproduce with a simple app created with `flutter create`? (If not, how big is your project, and is it open source?)

**Comment by DanTup** [comment]
Assuming it's the analysis server snapshot, if you add set the `"dart.analyzerObservatoryPort"` setting in your VS Code config to a free port and then restart VS / run the `Reload Window` command then open `localhost:(port)/` in your browser, you should be able to access Observatory for the analysis server where you should be able to get [some information about what's using the memory](https://dart-lang.github.io/observatory/allocation-profile.html) (let me know if you're not familiar with this and I can add some notes/screenshots to help).

(cc @bwilkerson @scheglov)

**Comment by peekpt** [comment]
@DanTup I'm trying to make it "crash" but I can't reproduce it again so far. I'm monitoring on the Analysis server and the logs are on. The code is private, but It happened with the sample app already. Trying make this go wild again...

**Comment by DanTup** [comment]
Ok; please do let us know if it happens again and what you manage to find. Be aware that if you leave those ports in your settings, it'll try and use the same port for all projects, so if you open multiple projects in separate VS Code windows, only one of them will bind. If this might happen, you can always set them at the Workspace level so each project has a different port.

**Comment by peekpt** [comment]
Yes the analyzing progress indicator never stops, that must be it... 
![](https://user-images.githubusercontent.com/222385/47081840-a6349980-d203-11e8-9b38-b020e04f651a.png)

It's very difficult to look at the server because it lags all the system out.

`[11:30:43 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6425","method":"edit.getFixes","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":802}}
[11:30:43 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6426","method":"edit.getAssists","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","length":0,"offset":802}}
[11:31:28 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6427","method":"analysis.updateContent","params":{"files":{"/Users/paulo/development/Projects/aadm/lib/main.dart":{"edits":[{"id":"","length":1,"offset":801,"replacement":""}],"type":"change"}}}}
[11:31:28 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6428","method":"completion.getSuggestions","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":956}}
[11:31:28 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6429","method":"completion.getSuggestions","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":956}}
[11:31:54 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6430","method":"edit.getAvailableRefactorings","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","length":0,"offset":767}}
[11:31:54 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6431","method":"edit.getFixes","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":767}}
[11:31:54 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6432","method":"edit.getAssists","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","length":0,"offset":767}}
[11:31:55 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6433","method":"analysis.getHover","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":1180}}
[11:31:58 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6434","method":"analysis.getHover","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":908}}
[11:31:58 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6435","method":"completion.getSuggestions","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":956}}
[11:32:10 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6436","method":"edit.getAvailableRefactorings","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","length":0,"offset":1082}}
[11:32:10 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6437","method":"edit.getFixes","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":1082}}
[11:32:10 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6438","method":"edit.getAssists","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","length":0,"offset":1082}}
[11:34:30 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6439","method":"edit.getAvailableRefactorings","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","length":0,"offset":717}}
[11:34:30 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6440","method":"edit.getFixes","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":717}}
[11:34:30 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6441","method":"edit.getAssists","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","length":0,"offset":717}}
[11:35:11 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6442","method":"edit.getAvailableRefactorings","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","length":0,"offset":719}}
[11:35:11 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6443","method":"edit.getFixes","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":719}}
[11:35:11 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6444","method":"edit.getAssists","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","length":0,"offset":719}}
[11:35:15 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6445","method":"analysis.getHover","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":930}}
[11:35:19 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6446","method":"analysis.getHover","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":746}}
[11:35:24 GMT+0100 (WEST)] [Analyzer] [Info] ==> {"id":"6447","method":"analysis.getHover","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":793}}
[11:40:19 GMT+0100 (WEST)] [Observatory] [Info] Observatory connection closed: 1006 ()
[11:40:41 GMT+0100 (WEST)] [FlutterRun] [Info] <== Lost connection to device.
[11:40:45 GMT+0100 (WEST)] [FlutterRun] [Info] <== [{"event":"app.stop","params":{"appId":"3f119462-c896-45de-b683-7176b0196e41"}}]
[11:40:45 GMT+0100 (WEST)] [Observatory] [Info] Process exited (sigterm)
[11:40:45 GMT+0100 (WEST)] [Observatory] [Info] Disconnect requested!
[11:40:45 GMT+0100 (WEST)] [Observatory] [Info] TERM: Going to terminate with SIGINT...
[11:40:45 GMT+0100 (WEST)] [Observatory] [Info] TERM: Removing all stored data...
[11:40:45 GMT+0100 (WEST)] [Observatory] [Info] TERM: Waiting for process to finish...
[11:40:46 GMT+0100 (WEST)] [Observatory] [Info] TERM: Disconnecting...
Wed Oct 17 2018 [11:42:36 GMT+0100 (WEST)] Log file ended
`

**Comment by peekpt** [comment]
Process sample:
https://paste.ofcode.org/3XCeLijjgaDUt4trRETDWq

**Comment by DanTup** [comment]
@peekpt Thanks! Are you able to share the full log from Dart Code? I'm curious to see what happened around the time of the last response from the Analyzer (which will have a `<==` instead of a `==>` in it).

It may also be worth opening the Observatory for the analyzer and checking in there (you could pause the isolate and see what it's doing). I'm not sure how useful the native dump is since the analyzer is all Dart code.

Also, if you can check the command line used to launch that process, we should confirm whether it is the analyzer and not one of the Flutter processes. It seems like the analyzer, but since you do have Flutter `daemon` and Flutter `run` going at the same time, we should confirm (something like `ps -ax | grep -v grep | grep dart` might get you that).

**Comment by peekpt** [comment]
[Dart-Code-Log-2018-09-02 19-45-25.txt.zip](https://github.com/flutter/flutter/files/2487180/Dart-Code-Log-2018-09-02.19-45-25.txt.zip)

**Comment by peekpt** [comment]
[Uploading logs_23155.zip…]()

**Comment by peekpt** [comment]
[logs_23155.zip](https://github.com/flutter/flutter/files/2487710/logs_23155.zip)

**Comment by peekpt** [comment]
![captura de ecra 2018-10-17 as 16 12 45](https://user-images.githubusercontent.com/222385/47097052-7481f980-d228-11e8-9c4d-27685da92d0e.png)
![captura de ecra 2018-10-17 as 16 20 13](https://user-images.githubusercontent.com/222385/47097104-954a4f00-d228-11e8-89ce-47535c5d6028.png)
![captura de ecra 2018-10-17 as 16 17 39](https://user-images.githubusercontent.com/222385/47097112-9aa79980-d228-11e8-82e3-6bd91c53c1ca.png)
![captura de ecra 2018-10-17 as 16 17 19](https://user-images.githubusercontent.com/222385/47097118-9da28a00-d228-11e8-9f8a-08186f0efe08.png)

**Comment by peekpt** [comment]
Can code a few minutes, then the rumble starts...
from analyzer logs I got:
(...)
[4:43:58 PM]: <== {"id":"1"}
[4:43:58 PM]: <== {"id":"2","result":{}}
[4:43:58 PM]: <== {"id":"3","result":{"assists":[]}}
[4:43:58 PM]: <== {"id":"5","result":{"assists":[]}}
[4:43:58 PM]: <== {"id":"8","result":{"assists":[]}}
[4:43:58 PM]: <== {"id":"11","result":{"assists":[]}}
[4:43:58 PM]: <== {"id":"4","result":{"fixes":[]}}
[4:43:58 PM]: <== {"id":"6","result":{"kinds":[]}}
[4:43:58 PM]: <== {"id":"7","result":{"fixes":[]}}
[4:43:58 PM]: <== {"id":"9","result":{"kinds":[]}}
[4:43:58 PM]: <== {"id":"10","result":{"fixes":[]}}
[4:43:58 PM]: <== {"event":"analysis.errors","params":{"file":"/Users/paulo/development/Projects/aadm/pubspec.yaml","errors":[]}}
[4:44:00 PM]: <== {"id":"12"}
[4:44:00 PM]: <== {"event":"server.status","params":{"analysis":{"isAnalyzing":true}}}
[4:44:03 PM]: <== {"event":"analysis.errors","params":{"file":"/Users/paulo/development/Projects/aadm/lib/api/blog.dart","errors":[]}}
[4:44:03 PM]: <== {"id":"13","result":{}}
[4:44:03 PM]: <== {"id":"14"}
[4:44:03 PM]: <== {"id":"15"}   ** <-- last reply**
[4:44:43 PM]: ==> {"id":"16","method":"analysis.getHover","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":2815}}
[4:44:43 PM]: ==> {"id":"17","method":"analysis.getHover","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":2152}}
[4:44:46 PM]: ==> {"id":"18","method":"analysis.getHover","params":{"file":"/Users/paulo/development/Projects/aadm/lib/main.dart","offset":2154}}
(...)

**Comment by peekpt** [comment]
Just an update: Updated Android Studio to 3.2.1 all the plugins etc, imported the project, same thing happens. Any news about this bug, I can't code :(

**Comment by scheglov** [comment]
I'd suggest trying the lates Dev channel Dart SDK.
We did several changes that we hope will fix the memory leak.

The problem is that it should be just *Dart* SDK, not the Flutter SDK.
And I don't know if it is possible to mix Flutter and some arbitrary Dart SDK :-(

**Comment by peekpt** [comment]
Thank you! IDK, it's automatic... Flutter automatically downloads the appropriate Dart build, I will try to switch to Flutter build master channel, it should have the latest dart apparently...

**Comment by peekpt** [comment]
Compiled the last dart , same thing (analyzing...) .
But I tried other project and it worked, then I closed open again the bug project and it briefly works, then breaks again... weird

$ ./dart --version
Dart VM version: 2.1.0-edge.5dfe3c8f94c9e880ad56d41316c0830817c96128 (Wed Oct 17 22:26:47 2018 +0000) on "macos_x64"

**Comment by peekpt** [comment]
It works, this is how I did it: found the dart's bin folder of the Flutter folder, move it temporarily to bin_old and symlinked the fresh compiled bin folder

**Comment by DanTup** [comment]
> The dart substitution worked, this is how I did it: found the dart's bin folder of the Flutter folder, move it temporarily to bin_old and symlinked the fresh compiled bin folder.

> I'm think there's something in the app code that breaks the analyzer, because I tried a different code , and the analyzer stops and the popups helpers worked

I'm not sure if this means it's working for you or you still have issues?

**Comment by peekpt** [comment]
Yes I'm still having issues, what worked was compiling the last dart and using it (bin folder symlinked)

**Comment by DanTup** [comment]
> Yes I'm still having issues, what worked was

Sorry, I'm still a little confused. Did your workaround (symlinking the bin folder) resolve the issues? Are you unblocked, or do you still have issues after doing that?

If the issues goes away, then it could be that the problem is fixed and an upcoming Dart roll into Flutter will fix things properly. If you're still seeing issues with the dev Dart SDK then there may still be issues we need to track down.

**Comment by peekpt** [comment]
@DanTup found the bug, can reproduce, going to upload the video, code related

**Comment by peekpt** [comment]
Note:
 I renamed all dart files to dartx so it won't affect the issue.
the only file bug.dart file on the lib dir
[bug](http://sendvid.com/luak5xji)

`
Future<List<int>> x() async {
  // if you type above it will break
}
`

**Comment by peekpt** [comment]
Just tried in a new fresh project and this line of codes breaks it

**Comment by peekpt** [comment]
@scheglov 
@DanTup  
`<foo<>>` breaks the analyser
paste this code and try to write anything above the line

**Comment by peekpt** [comment]
`<x<` also breaks

**Comment by scheglov** [comment]
Thank you, I can reproduce the problem on `foo Future<List<int>> bar() async {}`.

**Comment by DanTup** [comment]
@peekpt Great; thanks for tracking this down! I'll close this in favour of the SDK issue since I don't think there's anything Flutter-specific in here.

**Comment by peekpt** [comment]
Well, the bug is 'fixed' with a workaround. But now when this fix is coming to Flutter dart? 

I suppose this bugfix will be submitted to channel master, but many 3d party plugins only works with the dev channel because of its dependencies. 


So my question is since this is a critical bug how users will manage this without using the master channel?
Thank you

**Comment by Hixie** [comment]
We roll master into dev continually (as fast as we can -- sometimes multiple times a day, sometimes we run into problems and then it takes a few days); dev rolls into beta every few weeks. HTH.

**Comment by peekpt** [comment]
Thank you!

**Comment by tungvn** [comment]
Have you anything news on this issue? Seem I have the same issue?

**Comment by DanTup** [comment]
@tungvn The issue is believed to be fixed in the Flutter `dev` (and `master`) channels. If you're hitting it in `beta`, please try switching to the `dev` channel to see if it goes away (if it does not - please let us know) and switching back on the next update of the `beta` channel.

**Comment by tungvn** [comment]
Hmm, I'm still working on the `beta` channel and the bug is not happening anymore.
Now I try and can't reproduce it. I will try to do it at the next time I see it again.

**Comment by zeromaro** [comment]
![image](https://user-images.githubusercontent.com/12739071/80440548-32646880-8933-11ea-8653-367ae65c333e.png)

```
[✓] Flutter (Channel master, v1.18.0-6.0.pre.26, on Mac OS X 10.15.4 19E287, locale en-HK)
    • Flutter version 1.18.0-6.0.pre.26 at /Users/mac/FlutterProject/flutter
    • Framework revision 846e8fe812 (13 days ago), 2020-04-15 03:21:14 +0200
    • Engine revision deef2663ac
    • Dart version 2.8.0 (build 2.8.0-dev.20.0 c9710e5059)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/mac/Library/Android/sdk
    • Platform android-29, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4.1, Build version 11E503a
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.44.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.9.1

[✓] Connected device (4 available)
    • ONEPLUS A6000 • 2b4e4866                                 • android-arm64  • Android 10 (API 29)
    • iPhone        • ec0322d52fa501231677a7f06a283badc2beb08b • ios            • iOS 13.3.1
    • Chrome        • chrome                                   • web-javascript • Google Chrome 81.0.4044.122
    • Web Server    • web-server                               • web-javascript • Flutter Tools

• No issues found!
```
Just started facing this issue. Happens all the time when I want to launch my simulator.

**Comment by DanTup** [comment]
@zeromaro please open a new issue since this is closed and quite old. Things that would be useful to include:

- The full command line/args of the Dart process using the memory (something like `ps -e | grep dart` should give this - you can match the pid up with activity monitor)
- Whether you can reproduce this in a simple new `flutter create` project or only a specific project
- If the Dart process using the memory is the analysis server, an instrumentation log (VS Code instructions [here](https://dartcode.org/docs/logging/#analyzer-instrumentation)) - note: this will include parts of your source files so if the project is sensitive don't share, but instead review them yourself to see if there are any obvious issues in there

**Comment by maykhid** [comment]
@zeromaro Have you found a solution to this issue? Or did you open a new issue as @DanTup said. I need to know, I'm currently facing a similar problem.

**Comment by DanTup** [comment]
@maykhid it doesn't look like a new issue has been opened, so if you can repro this, please file one and include the info from my comment above (and cc me in it). Thanks!

**Comment by zeromaro** [comment]
I got quite busy recently - didnt file for external issue. I just used Android device for a couple of days or so. Then switched back to iOS - dart spike happens only maybe 2seconds. I have no idea what happened.

**Comment by zeromaro** [comment]
Just has this issue again, stayed on latest `master`. It started to reproduce, once I've upgraded my flutter.
Afterwards I have moved to `dev` - issue seems to be gone.

`Channel dev, 1.18.0-10.0.pre,`

**Comment by DanTup** [comment]
Are you able to open a new issue with the info mentioned in https://github.com/flutter/flutter/issues/23155#issuecomment-620454125?
